src/DynamoCoreWpf/Views/Preview/Watch3DView.xaml.cs

```diff
@@ -857,265 +890,125 @@ public void RenderDrawables(VisualizationEventArgs e)
-                }
-                else
+                var m = rp.Mesh;
+                if (m.Positions.Any())
                 {
-                    mesh.Colors.Add(ca);
-                    mesh.Colors.Add(cc);
-                    mesh.Colors.Add(cb); 
-                }
+                    // Pick a mesh to use to store the data. Selected geometry
+                    // goes into the selected mesh. Geometry with
+                    // colors goes into the per vertex mesh. Everything else
+                    // goes into the plain mesh.
 
-                color_idx += 12;
-                pt_idx += 3;
-            }
+                    var meshSet = rp.IsSelected ? 
+                        meshSelected :
+                        rp.RequiresPerVertexColoration ? perVertexMesh : mesh;
 
-            if (mesh.Indices.Count > 0)
-            {
-                MeshCount++;
-            }
-        }
+                    var idxCount = meshSet.Positions.Count;
 
-        private static Color4 GetColor(IRenderPackage p, int color_idx)
-        {
-            var color = new Color4(1,1,1,1);
+                    meshSet.Positions.AddRange(m.Positions);
 
-            if (color_idx <= p.TriangleVertexColors.Count-3)
-            {
-                color = new Color4(
-                (float)(p.TriangleVertexColors[color_idx] / 255.0),
-                (float)(p.TriangleVertexColors[color_idx + 1] / 255.0),
-                (float)(p.TriangleVertexColors[color_idx + 2] / 255.0),
-                (float)(p.TriangleVertexColors[color_idx + 3] / 255.0));
-            }
-           
-            return color;
-        }
+                    meshSet.Colors.AddRange(m.Colors);
+                    meshSet.Normals.AddRange(m.Normals);
+                    meshSet.TextureCoordinates.AddRange(m.TextureCoordinates);
+                    meshSet.Indices.AddRange(m.Indices.Select(i => i + idxCount));
 
-        private static Vector3 GetVertex(List<double> p, int i)
-        {
-            var x = (float)p[i];
-            var y = (float)p[i + 1];
-            var z = (float)p[i + 2];
+                    if (rp.DisplayLabels)
+                    {
+                        var pt = meshSet.Positions[idxCount];
+                        text.TextInfo.Add(new TextInfo(HelixRenderPackage.CleanTag(rp.Description), new Vector3(pt.X + 0.025f, pt.Y + 0.025f, pt.Z + 0.025f)));
+                    }
 
-            // DirectX convention - Y Up
-            var new_point = new Vector3(x, z, -y);
-            return new_point;
+                    MeshCount++;
+                }
+            }
         }
 
-        private string CleanTag(string tag)
+        private void SendGraphicsToView(
+            PointGeometry3D points,
+            LineGeometry3D lines,
+            LineGeometry3D linesSelected,
+            MeshGeometry3D mesh,
+            MeshGeometry3D meshSelected,
+            MeshGeometry3D perVertexMesh,
+            BillboardText3D text)
```

Comment:
Again, turn all of these parameters into a class so these method is readable.  The overhead is trivial.